seqr/views/apis/staff_api_tests.py

```diff
@@ -531,3 +606,87 @@ def test_anvil_export(self, mock_export_multiple_files):
+
+    @mock.patch('seqr.views.apis.staff_api.file_iter')
+    @mock.patch('seqr.views.apis.staff_api.logger')
+    def test_upload_qc_pipeline_output(self, mock_logger, mock_file_iter):
+        url = reverse(upload_qc_pipeline_output,)
+        _check_login(self, url)
+
+        # Test no dataset type error
+        mock_file_iter.return_value = SAMPLE_QC_DATA_NO_DATA_TYPE
+        response = self.client.post(url, content_type='application/json',
+                data=json.dumps({'file': 'gs://seqr-datasets/v02/GRCh38/RDG_WES_Broad_Internal/v15/sample_qc/final_output/seqr_sample_qc.tsv'}))
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.reason_phrase, 'No dataset type detected')
+
+        # Test multiple dataset types error
+        mock_file_iter.return_value = SAMPLE_QC_DATA_MORE_DATA_TYPE
+        response = self.client.post(url, content_type='application/json',
+                data=json.dumps({'file': 'gs://seqr-datasets/v02/GRCh38/RDG_WES_Broad_Internal/v15/sample_qc/final_output/seqr_sample_qc.tsv'}))
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.reason_phrase, 'Multiple dataset types detected: wes ,wgs')
+
+        # Test unexpected data type error
+        mock_file_iter.return_value = SAMPLE_QC_DATA_UNEXPECTED_DATA_TYPE
+        response = self.client.post(url, content_type='application/json',
+                data=json.dumps({'file': 'gs://seqr-datasets/v02/GRCh38/RDG_WES_Broad_Internal/v15/sample_qc/final_output/seqr_sample_qc.tsv'}))
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.reason_phrase, 'Unexpected dataset type detected: "unknown" (should be "exome" or "genome")')
+
+        # Test normal functions
+        mock_file_iter.return_value = SAMPLE_QC_DATA
+        mock_logger.reset_mock()
+        response = self.client.post(url, content_type='application/json',
+                data=json.dumps({'file': 'gs://seqr-datasets/v02/GRCh38/RDG_WES_Broad_Internal/v15/sample_qc/final_output/seqr_sample_qc.tsv'}))
+        self.assertEqual(response.status_code, 200)
+        response_json = response.json()
+        self.assertListEqual(response_json.keys(), ['info', 'errors', 'warnings'])
+        self.assertListEqual(response_json['info'], [
+            u'Parsed 4 exome samples',
+            u'Found and updated matching seqr individuals for 3 samples'
+        ])
+        self.assertListEqual(response_json['warnings'], [
+            u'The following 1 samples were skipped: 03133B_2',
+            u'The following filter flags have no known corresponding value and were not saved: coverage_exome',
+            u'The following population platform filters have no known corresponding value and were not saved: n_deletion, n_insertion, n_snp, r_insertion_deletion'
```

Comment:
This is similarly bad - you need to add back in the `sample_qc.*` columns to the mock raw data (`sample_qc. n_deletion`, `sample_qc. n_insertion`, etc) and then add an intentionally fake value into the  qc_metrics_filters column to check here

Reply:
Updated